backend/courses/views.py

```diff
@@ -376,3 +384,122 @@ def get_queryset(self):
+        res["pending"] = [model_to_dict(friendship) for friendship in pending_friendships]
+        res["rejected"] = [
+            model_to_dict(friendship) for friendship in rejected_friendships
+        ]  # unsure if we should include this
+        res["message"] = "Friendships retrieved successfully."
+        return JsonResponse(res)
+                
+    def post(self, request):
+        res = {}
+        # send friendship request to passed in recipient
+        sender = request.user
+        recipient = get_object_or_404(User, id=request.recipient_id)
+
+        if not recipient:
+            res["message"] = "Recipient does not exist."
+            return JsonResponse(res, status=400)
+
+        # if friend request already exists but was rejected, we can reinstate it:
+        friend_req = self.queryset.get(sender=sender, recipient=recipient, status=Friendship.FriendshipStatus.REJECTED)
+        if not friend_req:
+            friend_req = self.queryset.get(sender=recipient, recipient=sender, status=Friendship.FriendshipStatus.REJECTED)
+        if friend_req:
+            friend_req.status = Friendship.FriendshipStatus.SENT
+            friendship.save()
+            res = model_to_dict(friendship)
+            res["message"] = "Friendship request sent successfully."
+        
+        if self.queryset.get(sender=sender, recipient=recipient, status=Friendship.FriendshipStatus.SENT).exists():
+            res["message"] = "Friendship request already exists."
+            return JsonResponse(res, status=status.HTTP_409_CONFLICT)
+
+        # if a request exists in the other direction, then we accept the friend request
+        friendship_request = self.queryset.get(sender=recipient, recipient=sender, status=Friendship.FriendshipStatus.SENT)
+        if friendship_request:
+            # accept friend request
+            friendship_request.setStatus("A")
+            friendship.save()
+            res = model_to_dict(friendship_request)
+            res["message"] = "Friendship request accepted."
+        else:
+            # create the friendship request
+            friendship = Friendship(sender=sender, recipient=recipient)
+            friendship.status = Friendship.FriendshipStatus.SENT
+            friendship.save()
+            res = model_to_dict(friendship)
+            res["message"] = "Friendship request sent successfully."
+
+        return JsonResponse(res, status=200)
+
+    def delete(self, request):
+        # either deletes a friendship or cancels a friendship request (depends on status)
+        res = {}
+        sender = request.user
+        recipient = get_object_or_404(User, id=request.recipient_id)
+
+        if not recipient:
+            res["message"] = "Recipient does not exist."
+            return JsonResponse(res, status=400)
+
+        # delete the friendship/request (see if it exists the other way around)
+        friendship = self.queryset.get(sender=sender, recipient=recipient)
```

Comment:
Might want to use the same code as suggested above to get any friendships (with any status / in any direction) between these two, and then case on the direction to determine whether the object should actually be deleted (i.e. when `existing_friendship.sender_id == request.user.id`, it should be deleted -- otherwise its status should be changed to `Friendship.Status.REJECTED`).

Reply:
I'm planning on handling rejecting friendship requests in the post route. The delete route will only be to delete friendship requests or remove friends.